### PR TITLE
zypper: Fix installation from paths/urls

### PR DESCRIPTION
**Pull request by agx**
Currently s.th. like:
- name: Install Ansible RPM from local filesystem
  zypper:
    name='/vagrant/rpms/ansible-2.0.0-0.git201511170838.9a5342f.devel.noarch.rpm'

fails like:

  TASK [foo_ansible : Install Ansible RPM from local filesystem] ****************
  fatal: [foo.local]: FAILED! => {"changed": false, "failed": true, "msg": "No provider of 'ansible' found.\n"}

This happens since the actual name element is being ovewritten in
get_package_state. To fix this keep the real path in an extra dictionary so we
can reference it when we need to install the package.

Signed-off-by: Guido Günther agx@sigxcpu.org

### CODE REVIEW

File: packaging/os/zypper.py [hunk 1]
```diff
@@ -135,8 +135,23 @@ def get_current_version(m, packages):
...
 
 
+def names_to_paths(names, paths):
+    """
+    Map package names to paths if these are URLs/local paths but
+    use the package name otherwise
+    """
+    packages = []
+    for p in names:
+        if paths.has_key(p):
```
Comment (alxgu) [inline packaging/os/zypper.py]: better use "in" then has_key() to be compatibel with python3
https://docs.python.org/3.1/whatsnew/3.0.html#builtins

**Comment by agx** [comment]
@alxgu fixed. I've also aded a second patch that makes latest also work. Since this might be more debatable I've split this out into a separate one.

**Comment by robynbergeron** [comment]
Hi @agx -- thanks for submitting this PR. 

@alxgu -- thanks for already reviewing this once :) -- but some additional info for your benefit:

Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please add a comment with "shipit" in the text, and we will flag it for inclusion.

If it fails these guidelines, please add a comment with "needs_revision" in the text, along with the needed changes, and we will flag it for revision.

**Comment by robynbergeron** [comment]
Putting this in community_review as @alxgu asked for revisions and @agx has made them -- just waiting for more feedback or the goahead to ship it at this point.

**Comment by alxgu** [comment]
The second patch is good, too. In /var/log/zypp/history is no difference between install and update.
shipit

**Comment by agx** [comment]
Can this be merged?

**Comment by agx** [comment]
Ping? We have integration tests in ansible now to verify this.

**Comment by agx** [comment]
Ping? I'd would have been nice to have this in 2.0 to unbreak this on SLES/OpenSUSE. it is written to not break existing things so can it go in, please?

**Comment by bcoca** [comment]
this was waiting for a 'shipit' from the community, 2.0 has already been released

**Comment by alxgu** [comment]
okay, once again
shipit

**Comment by agx** [comment]
It has a shipit now, can it be merged?

**Comment by profhase** [comment]
Could you please merge this bugfix?

**Comment by agx** [comment]
Can this be merged. It has a shipit in https://github.com/ansible/ansible-modules-extras/pull/1256#issuecomment-160670455 since november.

**Comment by gregdek** [comment]
Thanks @agx for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @agx. To the current maintainers, @robinro please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by robinro** [comment]
@agx 
#2108 should also fix this. Please have a look and test it and let me know whether you are ok with using that instead of this PR.

**Comment by agx** [comment]
@robinro After five monthes i've totally lost focus on this one. I trust you that your changes also fix the issue.

**Pull request closed without merging**